### ISSUE TITLE
test_add_udt_to_another_data_types: Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#15530]
Although the symptom looks identical to #5336 I'm opening a new issue since it seems like we have a recent regression.

@kbr-scylla wrote:
> I've just got a very similar failure on https://github.com/scylladb/scylladb/pull/15331:
>
> https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3711/testReport/junit/user_types_test/TestUserTypes/Tests___Sanity_Tests___test_add_udt_to_another_data_types/
> ```
> cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="Invalid tuple literal for entity_
> ```

@Deexie hit this in https://github.com/scylladb/scylladb/pull/15506
> https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3843/

I just saw this in https://github.com/scylladb/scylladb/pull/15480#issuecomment-1732243520
> https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3886/testReport/junit/user_types_test/TestUserTypes/Tests___Sanity_Tests___test_add_udt_to_another_data_types/
> E   cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>"

**Comment by bhalevy** [comment]
let's determine first if the issue is with the product or with the test.

**Comment by bhalevy** [comment]
@nyh was there an recent change in this area on scylla's side?

**Comment by nyh** [comment]
@avikivity and @cvybhu have been changing related area of the CQL expression code. But if this failure isn't consistent (i.e., flaky. Is it?) perhaps a more likely possibility is that it's a schema change race - the request is done immediately after an ALTER TYPE and maybe it isn't aware of the change. It's not impossible that we had a regression somewhere - we could bisect. I can't think of  any specific change that I suspect.

**Comment by bhalevy** [comment]
@tgrabiec please confirm schema change sync issue

**Comment by nyh** [comment]
> @tgrabiec please confirm schema change sync issue

@bhalevy is this test failure flaky, or it recently started to fail consistently?

**Comment by bhalevy** [comment]
> > @tgrabiec please confirm schema change sync issue
> 
> @bhalevy is this test failure flaky, or it recently started to fail consistently?

It was flaky in the past and that issue was closed with no explicit fix (See https://github.com/scylladb/scylladb/issues/5336#issuecomment-570174818), but it became more flaky again recently so  I opened this issue rather than reopening #5336

**Comment by kbr-scylla** [comment]
Another occurrence
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3908/testReport/junit/user_types_test/TestUserTypes/Tests___Sanity_Tests___test_add_udt_to_another_data_types/

**Comment by cvybhu** [comment]
Usually the driver should wait for schema agreement after any `ALTER *` statements. It should wait until all nodes report the same schema version. Rust driver does this automatically, but I'm not sure about other drivers.

@avelanarius do you know if the Python driver automatically waits for schema agreement after `ALTER` queries? Or should we do it manually.

**Comment by cvybhu** [comment]
It looks like the default timeout for schema agreement in Python driver is 10 seconds. That might not be enough for slow clusters in dtest runners:

`max_schema_agreement_wait:`

![image](https://github.com/scylladb/scylladb/assets/36861778/7bc020a8-c07b-47b8-a44f-60a9dd5d9715)

**Comment by kbr-scylla** [comment]
We could modify the test to enable DEBUG logging in the Python driver and see if the schema agreement wait times out.

If it does though, I don't believe it is a matter of increasing the timeout. Why did the previous timeout suffice until now, but no longer does? It could be that for some reason, we never reach this agreement.

We could of course increase the timeout and see if it helps.

**Comment by cvybhu** [comment]
I had a look at Python driver's code, and I noticed two things:
* I don't see any logic that would await schema agreement after `ALTER` queries automatically. Maybe we have to call `refresh_metadata()` manually after each ALTER? Or maybe the logic is there but I missed it.
* Python driver waits for schema agreement using only the control connection. It asks the control connection shard whether the local schema version and peers schema versions are the same, but all of this information comes from one node. This could lead to races, for example:
    * Driver has control connection to node A
    * Driver sends an `ALTER` query to node B
    * Driver asks node A for schema versions, they are all the same because information about the `ALTER` query hasn't propagated to node A yet.
    * Driver things that schema agreement has been reached, but node A still doesn't know about the schema change

  Rust driver does it differently, it queries the local schema version on all available connections and checks whether the local schema versions are the same on all shards. This makes such race conditions impossible.

I suspect that the issue is caused by Python driver's handling of schema agreement, but I'm not familiar with this codebase. I think it would be good to get an opinion from @Lorak-mmk or @avelanarius here. Maybe there are some known quirks in this area.

**Comment by avelanarius** [comment]
> * I don't see any logic that would await schema agreement after `ALTER` queries automatically. Maybe we have to call `refresh_metadata()` manually after each ALTER? Or maybe the logic is there but I missed it.

It does it here: https://github.com/scylladb/python-driver/blob/d735957e3a7b7178f4fff26f6dbae588e58cf314/cassandra/cluster.py#L4728-L4735

**Comment by nyh** [comment]
> It looks like the default timeout for schema agreement in Python driver is 10 seconds. That might not be enough for slow clusters in dtest runners:
> 
> `max_schema_agreement_wait:`
> 
> ![image](https://user-images.githubusercontent.com/36861778/270361010-7bc020a8-c07b-47b8-a44f-60a9dd5d9715.png)

Note: dtest has in dtest_setup.py ``max_schema_agreement_wait=60`. This should be (famous last words...) more than enough. I suspect we don't wait for schema agreement at all...

**Comment by nyh** [comment]
> I had a look at Python driver's code, and I noticed two things:
> 
>     * I don't see any logic that would await schema agreement after `ALTER` queries automatically. Maybe we have to call `refresh_metadata()` manually after each ALTER? Or maybe the logic is there but I missed it.

The way this is done automatically is as follows:
1. As a response to a schema-modifying request, Scylla replies with a special response code defined in the CQL protocol -  "Schema_change" (0x5).
2. Upon receiving a schema_change response ("event"), the driver should wait for the schema agreement.

As far as I can tell, Scylla does return SCHEMA_CHANGE on "ALTER TABLE" - see cql3/statements/alter_table_statement.cc,
and the Python driver does have code to handle this. But I didn't look too carefully, and didn't try to enable logging to see what really happens in practice.

>     * Python driver waits for schema agreement using only the control connection. It asks the control connection shard whether the local schema version and peers schema versions are the same, but all of this information comes from one node. This could lead to races, for example:
>       
>       * Driver has control connection to node A
>       * Driver sends an `ALTER` query to node B
>       * Driver asks node A for schema versions, they are all the same because information about the `ALTER` query hasn't propagated to node A yet.

If this is what the driver does, it's indeed looks like a bug :-( The driver knows that it just sent a specific connection to node B a request which node B told it on this connection that the schema to change. It should ask node B on the same connection - not anyone else - if it thinks the schema is agreed. If this is not what the code does, it's a bug :-(

>       * Driver thinks that schema agreement has been reached, but node A still doesn't know about the schema change
>       
>       Rust driver does it differently, it queries the local schema version on all available connections and checks whether the local schema versions are the same on all shards. This makes such race conditions impossible.

This makes more sense and probably converges faster, but sounds like more work for the client. Does the rust driver always open connections to all nodes in the cluster, right on the first request?

**Comment by cvybhu** [comment]
> It does it here: https://github.com/scylladb/python-driver/blob/d735957e3a7b7178f4fff26f6dbae588e58cf314/cassandra/cluster.py#L4728-L4735
> A query to fetch schema versions from system.local and system.peers if performed on the same connection that we have received RESULT_KIND_SCHEMA_CHANGE from. I don't think the issue you described can happen.

Thanks, you're right, that refutes my complaints. The race condition won't happen if we query the same connection where the ALTER was sent to. It looks like the Python driver does the right thing.

**Comment by cvybhu** [comment]
I enabled debug logging for python driver, run the and collected the logs.

```diff
diff --git a/conftest.py b/conftest.py
index a2373541..28d26443 100644
--- a/conftest.py
+++ b/conftest.py
@@ -194,7 +194,7 @@ def fixture_logging_setup(request, record_property, split_info):
         error_logger.addHandler(logging_plugin.log_file_handler)
         log_per_process_data['error_logger'] = error_logger
 
-    logging.getLogger("cassandra").setLevel(logging.INFO)
+    logging.getLogger("cassandra").setLevel(logging.DEBUG)
     logging.getLogger("boto3").setLevel(logging.INFO)
     logging.getLogger("botocore").setLevel(logging.INFO)
     logging.getLogger("s3transfer").setLevel(logging.INFO)
```

Here are the logs from a successful run:

[debug-level-logs-schema-change.txt](https://github.com/scylladb/scylladb/files/12730338/debug-level-logs-schema-change.txt)

Python driver does wait for schema agreement (All the lines with `Waiting for schema agreement`).

The only thing that surprised me is that there's only one message like this:
`Refreshing schema in response to schema change`
I would expect one such message for every schema change (?)

**Comment by cvybhu** [comment]
> We could modify the test to enable DEBUG logging in the Python driver and see if the schema agreement wait times out.

This sounds like a good way forward.
But is it possible to enable `DEBUG` logging for cassandra in this single dtest test case? I was able to do it globally, but I don't see how to do this for a single test.

**Comment by cvybhu** [comment]
Alright I hacked something together and opened a PR. The next time the CI fails we should be able to see logs about schema agreement.

**Comment by cvybhu** [comment]
Okay we got the same failure in e.g https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/5021/

There is no schema agreement timeout, but it looks like the test executes the query while the driver is still waiting for schema agreement:
```
<cassandra.cluster.ControlConnection object at 0x7f53d2700090>>, (), (('target_type', 'TABLE'), ('change_type', 'UPDATED'), ('keyspace', 'abcinfo'), ('table', 'entity')))
10:51:49,910 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TABLE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'table': 'entity_by_unique_id'}, stream_id=-1, trace_id=None)>
10:51:49,910 787     cassandra.cluster              DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Ignoring schedule_unique for already-scheduled task: (<bound method ControlConnection.refresh_schema of <cassandra.cluster.ControlConnection object at 0x7f53d2700090>>, (), (('target_type', 'TABLE'), ('change_type', 'UPDATED'), ('keyspace', 'abcinfo'), ('table', 'entity_by_unique_id')))
10:51:49,911 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TABLE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'table': 'entity_by_type'}, stream_id=-1, trace_id=None)>
10:51:49,911 787     cassandra.cluster              DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Ignoring schedule_unique for already-scheduled task: (<bound method ControlConnection.refresh_schema of <cassandra.cluster.ControlConnection object at 0x7f53d2700090>>, (), (('target_type', 'TABLE'), ('change_type', 'UPDATED'), ('keyspace', 'abcinfo'), ('table', 'entity_by_type')))
10:51:49,987 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: Refreshing schema in response to schema change. {'target_type': 'TYPE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'type': 'user_refs'}
10:51:49,988 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Waiting for schema agreement
10:51:49,989 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:50,190 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:50,391 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:50,592 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:50,793 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:50,994 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas match
10:51:50,994 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Waiting for schema agreement
10:51:50,995 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas match
10:51:50,996 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Waiting for schema agreement
10:51:51,093 787     errors                         ERROR    conftest.py         :225  | test_add_udt_to_another_data_types: test failed: 
self = <user_types_test.TestUserTypes object at 0x7f53d9754050>

    def test_add_udt_to_another_data_types(self):
        logging.getLogger("cassandra").setLevel(logging.DEBUG)
        try:
>           self.internal_add_udt_to_another_data_types()

user_types_test.py:1008: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
user_types_test.py:1124: in internal_add_udt_to_another_data_types
    insert_into_entity(rows=10, altered_type=True)
user_types_test.py:1083: in insert_into_entity
    session.execute(stmt)
cassandra/cluster.py:2700: in cassandra.cluster.Session.execute
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>"
```

There's `Waiting for schema agreement` and immediately after the failure happens. At the first glance it looks like the driver starts the schema agreement process, but the query doesn't wait for it.

Here's the full log: [with-debug-logs.log](https://github.com/scylladb/scylladb/files/12803386/with-debug-logs.log)

**Comment by nyh** [comment]
> There's `Waiting for schema agreement` and immediately after the failure happens. At the first glance it looks like the driver starts the schema agreement process, but the query doesn't wait for it.

Good catch, this confirms my guess in https://github.com/scylladb/scylladb/issues/15530#issuecomment-1735831569 that the problem is not the amount of wait (which is already 60 seconds), but the fact that the driver doesn't wait for it to finish. 

So sounds like a driver bug, but tere's nothing smart I can say about any details. I wonder if this test failure can be reproduced locally (even if at low probability) - it will help debug it and more importantly, test a fix.

**Comment by fruch** [comment]
elastic data shows it happen almost on a daily basis:
https://70f106c98484448dbc4705050eb3f7e9.us-east-1.aws.found.io:9243/goto/38708670-63d0-11ee-bf28-5927cc3d3ce3

**Comment by bhalevy** [comment]
Cc @avikivity :point_up:

**Comment by fruch** [comment]
What we are seeing is a bit more complex than one alter, 
we have 3 schema changes CQL command one after the other:

```
        logger.info('Create priority_refs type')
        session.execute('create type if not exists priority_refs (priority int, description set<int>)')

        logger.info('Alter obs_entity type')
        session.execute('alter type obs_entity add priority_refs frozen<priority_refs>')

        logger.info('Alter user_refs type')
        session.execute('alter type user_refs add priority_refs frozen<priority_refs>')
```

followed by 10 insert commands.

driver is handling echa schama change one after the other, but there's nothing blocking new query to be executed at that time.
I don't any code preventing queries.

no wonder we have in dtest code like `change_schema_safely` and `wait_for_schema_agreement` 
and `session.cluster.control_connection.wait_for_schema_agreement()` sprinkles just for safety all over the place.

@avelanarius @Lorak-mmk, any thoughts ? or i'm missing something ?

**Comment by Lorak-mmk** [comment]
IIUC those are synchronous queries and each is schema altering, so each of them should finish only after schema is in agreement, @avelanarius linked the code few comments above

**Comment by avikivity** [comment]
@fruch those session.execute calls are synchronous. Yes, the call to refresh_schema_and_set_result is asynchronous, but the wrapper (session.execute) should wait for it.

**Comment by Lorak-mmk** [comment]
> 
> There's `Waiting for schema agreement` and immediately after the failure happens. At the first glance it looks like the driver starts the schema agreement process, but the query doesn't wait for it.
> 

I don't think that's the case. As you can see in the logs, the driver got feedback about schema change, started schema agreement wait, and finished it (there is a `Schemas match` entry). That means schema was in agreement after the query.
It would be good to have a log before those lines:
```
insert_into_entity(rows=10, altered_type=True)
insert_into_entity_rel(rows=10, altered_type=True)
```
so that we know if schema was agreed before we begin insertion, but I suspect that it is and a problem is different.

**Comment by bhalevy** [comment]
Does the cliwnf wait for schema agreement on all shards?

**Comment by avikivity** [comment]
The logs show we do wait:

```
10:51:47,555 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Opening new connection to 127.0.6.1:9042
10:51:47,556 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: Error trying to reconnect control connection: NoHostAvailable('Unable to connect to any servers', {'127.0.6.1:9042': ConnectionRefusedError(111, "Tried connecting to [('127.0.6.1', 9042)]. Last error: Connection refused")})
10:51:48,269 787     /jenkins/workspace/scylla-master/gating-dtest-release/scylla-dtest/user_types_test.py INFO     user_types_test.py  :1115 | test_add_udt_to_another_data_types: Create priority_refs type
```


A type is created.


```
10:51:48,271 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TYPE', 'change_type': 'CREATED', 'keyspace': 'abcinfo', 'type': 'priority_refs'}, stream_id=-1, trace_id=None)>
10:51:48,280 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: Refreshing schema in response to schema change. {'target_type': 'TYPE', 'change_type': 'CREATED', 'keyspace': 'abcinfo', 'type': 'priority_refs'}
10:51:48,280 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Waiting for schema agreement
10:51:48,281 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:48,482 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:48,683 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:48,883 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas match
10:51:48,884 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Waiting for schema agreement
```

sync

```
10:51:48,884 787     /jenkins/workspace/scylla-master/gating-dtest-release/scylla-dtest/user_types_test.py INFO     user_types_test.py  :1118 | test_add_udt_to_another_data_types: Alter obs_entity type
```

ALTER


```
10:51:48,885 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas match
```

(probably the response to the event from the previous ALTER)


```
10:51:48,891 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TYPE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'type': 'obs_entity'}, stream_id=-1, trace_id=None)>
10:51:48,893 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TABLE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'table': 'entity_rel'}, stream_id=-1, trace_id=None)>
10:51:48,893 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TABLE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'table': 'entity'}, stream_id=-1, trace_id=None)>
10:51:48,893 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TABLE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'table': 'entity_by_unique_id'}, stream_id=-1, trace_id=None)>
10:51:48,893 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TABLE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'table': 'entity_by_type'}, stream_id=-1, trace_id=None)>
10:51:48,895 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: Refreshing schema in response to schema change. {'target_type': 'TYPE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'type': 'obs_entity'}
```

Waiting for agreement

```
10:51:48,895 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Waiting for schema agreement
10:51:48,896 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:49,097 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:49,298 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:49,499 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:49,700 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:49,901 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas match
10:51:49,901 787     /jenkins/workspace/scylla-master/gating-dte
```

matched

```
st-release/scylla-dtest/user_types_test.py INFO     user_types_test.py  :1121 | test_add_udt_to_another_data_types: Alter user_refs type
```

Second alter

```
10:51:49,908 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TYPE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'type': 'user_refs'}, stream_id=-1, trace_id=None)>
10:51:49,908 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TYPE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'type': 'obs_entity'}, stream_id=-1, trace_id=None)>
10:51:49,908 787     cassandra.cluster              DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Ignoring schedule_unique for already-scheduled task: (<bound method ControlConnection.refresh_schema of <cassandra.cluster.ControlConnection object at 0x7f53d2700090>>, (), (('target_type', 'TYPE'), ('change_type', 'UPDATED'), ('keyspace', 'abcinfo'), ('type', 'obs_entity')))
10:51:49,910 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TABLE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'table': 'entity_rel'}, stream_id=-1, trace_id=None)>
10:51:49,910 787     cassandra.cluster              DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Ignoring schedule_unique for already-scheduled task: (<bound method ControlConnection.refresh_schema of <cassandra.cluster.ControlConnection object at 0x7f53d2700090>>, (), (('target_type', 'TABLE'), ('change_type', 'UPDATED'), ('keyspace', 'abcinfo'), ('table', 'entity_rel')))
10:51:49,910 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TABLE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'table': 'entity'}, stream_id=-1, trace_id=None)>
10:51:49,910 787     cassandra.cluster              DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Ignoring schedule_unique for already-scheduled task: (<bound method ControlConnection.refresh_schema of <cassandra.cluster.ControlConnection object at 0x7f53d2700090>>, (), (('target_type', 'TABLE'), ('change_type', 'UPDATED'), ('keyspace', 'abcinfo'), ('table', 'entity')))
10:51:49,910 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TABLE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'table': 'entity_by_unique_id'}, stream_id=-1, trace_id=None)>
10:51:49,910 787     cassandra.cluster              DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Ignoring schedule_unique for already-scheduled task: (<bound method ControlConnection.refresh_schema of <cassandra.cluster.ControlConnection object at 0x7f53d2700090>>, (), (('target_type', 'TABLE'), ('change_type', 'UPDATED'), ('keyspace', 'abcinfo'), ('table', 'entity_by_unique_id')))
10:51:49,911 787     cassandra.connection           DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Message pushed from server: <EventMessage(event_type='SCHEMA_CHANGE', event_args={'target_type': 'TABLE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'table': 'entity_by_type'}, stream_id=-1, trace_id=None)>
10:51:49,911 787     cassandra.cluster              DEBUG    libevreactor.py     :370  | test_add_udt_to_another_data_types: Ignoring schedule_unique for already-scheduled task: (<bound method ControlConnection.refresh_schema of <cassandra.cluster.ControlConnection object at 0x7f53d2700090>>, (), (('target_type', 'TABLE'), ('change_type', 'UPDATED'), ('keyspace', 'abcinfo'), ('table', 'entity_by_type')))
10:51:49,987 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: Refreshing schema in response to schema change. {'target_type': 'TYPE', 'change_type': 'UPDATED', 'keyspace': 'abcinfo', 'type': 'user_refs'}
10:51:49,988 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Waiting for schema agreement
10:51:49,989 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:50,190 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:50,391 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:50,592 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:50,793 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas mismatched, trying again
10:51:50,994 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas match
10:51:50,994 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Waiting for schema agreement
10:51:50,995 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Schemas match
10:51:50,996 787     cassandra.cluster              DEBUG    thread.py           :58   | test_add_udt_to_another_data_types: [control connection] Waiting for schema agreement
```

What happened here?! Why didn't we complete waiting?

```
10:51:51,093 787     errors                         ERROR    conftest.py         :225  | test_add_udt_to_another_data_types: test failed: 
self = <user_types_test.TestUserTypes object at 0x7f53d9754050>

    def test_add_udt_to_another_data_types(self):
        logging.getLogger("cassandra").setLevel(logging.DEBUG)
        try:
>           self.internal_add_udt_to_another_data_types()

user_types_test.py:1008: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
user_types_test.py:1124: in internal_add_udt_to_another_data_types
    insert_into_entity(rows=10, altered_type=True)
user_types_test.py:1083: in insert_into_entity
    session.execute(stmt)
cassandra/cluster.py:2700: in cassandra.cluster.Session.execute
    ???
```

**Comment by avikivity** [comment]
> Does the client wait for schema agreement on all shards?

Schema is per-node. We need to check that we update the peers table only after all shards agree. I don't know how that happens, but hard to see a race there.

**Comment by avikivity** [comment]
So it looks like we did wait.

Possible probems:
 - did not reach all shards
 - did not invalidate prepared statements

**Comment by avikivity** [comment]
The test doesn't use prepared statements, so not invalidating them doesn't explain anything.

**Comment by avikivity** [comment]
@Lorak-mmk is it using preloaded_results != None? my Python isn't good enough to tell.

**Comment by Lorak-mmk** [comment]
> @Lorak-mmk is it using preloaded_results != None? my Python isn't good enough to tell.

In case of `RESULT_KIND_SCHEMA_CHANGE`? I'm pretty sure it's `None`

**Comment by avikivity** [comment]
Ok. Hard to see where it's broken. Does it reproduce locally? I'm afraid of dtest.

**Comment by avelanarius** [comment]
> Ok. Hard to see where it's broken. Does it reproduce locally? I'm afraid of dtest.

I tried it locally for a couple 100s of repeats (~2h) and I couldn't reproduce it locally.

**Comment by avikivity** [comment]
Maybe we should add a side load to juggle the timings. Perhaps running N copies in parallel (where N > vcpus_in_machine).

**Comment by tgrabiec** [comment]
On Sun, Oct 8, 2023 at 5:01 PM Avi Kivity ***@***.***> wrote:

> Does the client wait for schema agreement on all shards?
>
> Schema is per-node. We need to check that we update the peers table only
> after all shards agree. I don't know how that happens, but hard to see a
> race there.
>

We announce the per-node schema version after do_merge_schema(), which
updates all shards.

**Comment by cvybhu** [comment]
Maybe it'd be possible to set up a jenkins runner similar to the one for the CI and `tcpdump` all traffic between the python driver and the cluster. Inspecting the individual requests would allow to find out whether the error is in the driver or the database.

The rust driver has a proxy that can be put in the middle between the driver and the database, but I'm not sure if it can be used for dumping all the messages. https://github.com/scylladb/scylla-rust-driver/tree/main/scylla-proxy

**Comment by avikivity** [comment]
Happened again: https://jenkins.scylladb.com/view/nexts/job/scylla-master/job/next/6633/testReport/junit/user_types_test/TestUserTypes/Tests___Sanity_Tests_RAFT___test_add_udt_to_another_data_types/

Maybe it's happening more with driver version 3.26.3?

**Comment by avikivity** [comment]
> Happened again: https://jenkins.scylladb.com/view/nexts/job/scylla-master/job/next/6633/testReport/junit/user_types_test/TestUserTypes/Tests___Sanity_Tests_RAFT___test_add_udt_to_another_data_types/
> 
> Maybe it's happening more with driver version 3.26.3?

Unlikely, the driver in the toolchain was updated in Oct 4. But likely dtest just downloads the most recent driver.

**Comment by fruch** [comment]
> > Happened again: https://jenkins.scylladb.com/view/nexts/job/scylla-master/job/next/6633/testReport/junit/user_types_test/TestUserTypes/Tests___Sanity_Tests_RAFT___test_add_udt_to_another_data_types/
> > Maybe it's happening more with driver version 3.26.3?
> 
> Unlikely, the driver in the toolchain was updated in Oct 4. But likely dtest just downloads the most recent driver.

dtest doesn't download latest, but we did switched to 3.26.3 in https://github.com/scylladb/scylla-dtest/pull/3608 (Oct 3)
and those issues are from before that

**Comment by avikivity** [comment]
Probably the best way is to bisect. But for that we need a reliable local reproducer.

**Comment by kbr-scylla** [comment]
Hmm, if you suspect it's schema change code -- could it be 3d4398d1b2b9bc2139718e49a806214171ee081f?
I'm not sure how, but it was a significant change to how schema changes are done (at least in Raft mode).

**Comment by kbr-scylla** [comment]
A simple way to revert the effect of https://github.com/scylladb/scylladb/commit/3d4398d1b2b9bc2139718e49a806214171ee081f without reverting the entire thing would be:
```diff
diff --git a/service/migration_manager.cc b/service/migration_manager.cc
index e289ff7ce6..5b7f491139 100644
--- a/service/migration_manager.cc
+++ b/service/migration_manager.cc
@@ -1008,7 +1008,7 @@ static void add_committed_by_group0_flag(std::vector<mutation>& schema, const gr
 
 // Returns a future on the local application of the schema
 future<> migration_manager::announce(std::vector<mutation> schema, group0_guard guard, std::string_view description) {
-    if (_feat.group0_schema_versioning) {
+    if (false) {
         schema.push_back(make_group0_schema_version_mutation(_storage_proxy.data_dictionary(), guard));
         add_committed_by_group0_flag(schema, guard);
     }
```

**Comment by kbr-scylla** [comment]
Looking at the schema change code -- it looks like it's updating types on all shards before returning.
```cpp
static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, sharded<db::system_keyspace>& sys_ks, std::vector<mutation> mutations, bool do_flush, bool reload)
{
...
    auto types_to_drop = co_await merge_types(proxy, std::move(old_types), std::move(new_types));
```

```cpp
static future<user_types_to_drop> merge_types(distributed<service::storage_proxy>& proxy, schema_result before, schema_result after)
{
    auto diff = diff_rows(before, after);

    // Create and update user types before any tables/views are created that potentially
    // use those types. Similarly, defer dropping until after tables/views that may use
    // some of these user types are dropped.

    co_await proxy.local().get_db().invoke_on_all([&] (replica::database& db) -> future<> {
        for (auto&& user_type : create_types(db, diff.created)) {
            db.find_keyspace(user_type->_keyspace).add_user_type(user_type);
            co_await db.get_notifier().create_user_type(user_type);
        }
        for (auto&& user_type : create_types(db, diff.altered)) {
            db.find_keyspace(user_type->_keyspace).add_user_type(user_type);
            co_await db.get_notifier().update_user_type(user_type);
        }
    });
```

and (`do_merge_schema` must finish locally before Scylla returns from the schema change request)

**Comment by avikivity** [comment]
Kudos for getting a reproducer. If it reproduces 1 in 100 runs (say), just run it 1000 times (with concurrency). We can then use bisect to look for the bad commit (we can try to prove that a commit 1 week older than the first report is okay, to have a good commit.

**Comment by kbr-scylla** [comment]
I applied the following diff to the reproducer:
```diff
diff --git a/src/main.rs b/src/main.rs
index 1c405de..a8ca9d9 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -181,6 +181,7 @@ async fn main() {
     let mut args = std::env::args();
     let _name = args.next().unwrap();
     let ip = args.next().unwrap();
+    for iter in 0..1000 {
     let mut ctrl_conn = setup_connection(&ip, None).await;
     let mut shards_conns: Vec<Connection> = Vec::new();
     for shard in 0..ctrl_conn.shard_count {
@@ -366,6 +367,8 @@ async fn main() {
             .await
     );
 
+    println!("Iter {}", iter);
+
     println!("Alter user_refs type");
     println!(
         "{:?}",
@@ -382,4 +385,5 @@ async fn main() {
 
     // assert_row_count(session=session, table_name="entity_rel", expected=rows_num,
     //                     consistency_level=ConsistencyLevel.QUORUM)
+    }
 }
```

it reproduced after 50 iterations

**Comment by avikivity** [comment]
That's good. So 1000 iterations (with concurrency) should be a reliable pass/fail test (provided we get at least 10 fails from 1000, not 1).

**Comment by kbr-scylla** [comment]
More details about the error can be obtained by modifying Scylla a bit:
```diff
@@ -528,6 +529,7 @@ tuple_constructor_test_assignment(const tuple_constructor& tc, data_dictionary::
         tuple_constructor_validate_assignable_to(tc, db, keyspace, schema_opt, receiver);
         return assignment_testable::test_result::WEAKLY_ASSIGNABLE;
     } catch (exceptions::invalid_request_exception& e) {
+        expr_logger.warn("tuple not assignable: {}", e);
         return assignment_testable::test_result::NOT_ASSIGNABLE;
     }
 }
```

now Scylla prints a warning before throwing:
```
WARN  2023-10-10 14:39:08,581 [shard 16:stat] cql_expression - tuple not assignable: exceptions::invalid_request_exception (Invalid tuple literal for entity_info[5]: too many elements. Type frozen<user_refs> expects 5 but got 6)
```

**Comment by Lorak-mmk** [comment]
- The issue probably doesn't happen with `--consistent-cluster-management=false` - despite a lot of tries I have not been able to reproduce it there.
- It's not a matter of some single shard being out of date. I modified the reproducer to always print results from all shard, and when the failure happens it happens on all of them:
```
Bound socket to: 0.0.0.0:12024
Created connection to shard 0
Bound socket to: 0.0.0.0:12025
Created connection to shard 1
Bound socket to: 0.0.0.0:12026
Created connection to shard 2
Bound socket to: 0.0.0.0:12027
Created connection to shard 3
Bound socket to: 0.0.0.0:12028
Created connection to shard 4
Bound socket to: 0.0.0.0:12029
Created connection to shard 5
Bound socket to: 0.0.0.0:12030
Created connection to shard 6
Bound socket to: 0.0.0.0:12031
Created connection to shard 7
Bound socket to: 0.0.0.0:12032
Created connection to shard 8
Bound socket to: 0.0.0.0:12033
Created connection to shard 9
Bound socket to: 0.0.0.0:12034
Created connection to shard 10
Bound socket to: 0.0.0.0:12035
Created connection to shard 11
Drop keyspace
Ok(SchemaChange(SchemaChange { event: KeyspaceChange { change_type: Dropped, keyspace_name: "ks" } }))
Create keyspace
Ok(SchemaChange(SchemaChange { event: KeyspaceChange { change_type: Created, keyspace_name: "ks" } }))
Control connection - use keyspace
Ok(SetKeyspace(SetKeyspace { keyspace_name: "ks" }))
Create user_refs type
Ok(SchemaChange(SchemaChange { event: TypeChange { change_type: Created, keyspace_name: "ks", type_name: "user_refs" } }))
Create obs_entity type
Ok(SchemaChange(SchemaChange { event: TypeChange { change_type: Created, keyspace_name: "ks", type_name: "obs_entity" } }))
Create other_entity type
Ok(SchemaChange(SchemaChange { event: TypeChange { change_type: Created, keyspace_name: "ks", type_name: "other_entity" } }))
Create entity table
Ok(SchemaChange(SchemaChange { event: TableChange { change_type: Created, keyspace_name: "ks", object_name: "entity" } }))
Create index on entity(service)
Ok(SchemaChange(SchemaChange { event: TableChange { change_type: Updated, keyspace_name: "ks", object_name: "entity" } }))
Create index on entity(type)
Ok(SchemaChange(SchemaChange { event: TableChange { change_type: Updated, keyspace_name: "ks", object_name: "entity" } }))
Create entity_by_type materialized view
Ok(SchemaChange(SchemaChange { event: TableChange { change_type: Created, keyspace_name: "ks", object_name: "entity_by_type" } }))
Create entity_by_unique_id materialized view
Ok(SchemaChange(SchemaChange { event: TableChange { change_type: Created, keyspace_name: "ks", object_name: "entity_by_unique_id" } }))
Create entity_rel table
Ok(SchemaChange(SchemaChange { event: TableChange { change_type: Created, keyspace_name: "ks", object_name: "entity_rel" } }))
Create index on entity_rel(dest_entity_id)
Ok(SchemaChange(SchemaChange { event: TableChange { change_type: Updated, keyspace_name: "ks", object_name: "entity_rel" } }))
Create index on entity_rel(service)
Ok(SchemaChange(SchemaChange { event: TableChange { change_type: Updated, keyspace_name: "ks", object_name: "entity_rel" } }))
Create index on entity_rel(rel_type)
Ok(SchemaChange(SchemaChange { event: TableChange { change_type: Updated, keyspace_name: "ks", object_name: "entity_rel" } }))
Inserted on shard 5
Inserted on shard 8
Inserted on shard 1
Inserted on shard 2
Inserted on shard 4
Inserted on shard 6
Inserted on shard 10
Inserted on shard 7
Inserted on shard 9
Inserted on shard 11
Inserted on shard 0
Inserted on shard 3
Inserted on shard 8
Inserted on shard 5
Inserted on shard 0
Inserted on shard 11
Inserted on shard 3
Inserted on shard 9
Inserted on shard 4
Inserted on shard 6
Inserted on shard 2
Inserted on shard 7
Inserted on shard 10
Inserted on shard 1
Create priority_refs type
Ok(SchemaChange(SchemaChange { event: TypeChange { change_type: Created, keyspace_name: "ks", type_name: "priority_refs" } }))
Alter obs_entity type
Ok(SchemaChange(SchemaChange { event: TypeChange { change_type: Updated, keyspace_name: "ks", type_name: "obs_entity" } }))
Alter user_refs type
Ok(SchemaChange(SchemaChange { event: TypeChange { change_type: Updated, keyspace_name: "ks", type_name: "user_refs" } }))
FAILED on shard 3: Error { error: Invalid, reason: "Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>" }
FAILED on shard 5: Error { error: Invalid, reason: "Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>" }
FAILED on shard 6: Error { error: Invalid, reason: "Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>" }
FAILED on shard 1: Error { error: Invalid, reason: "Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>" }
FAILED on shard 2: Error { error: Invalid, reason: "Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>" }
FAILED on shard 4: Error { error: Invalid, reason: "Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>" }
FAILED on shard 8: Error { error: Invalid, reason: "Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>" }
FAILED on shard 7: Error { error: Invalid, reason: "Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>" }
FAILED on shard 9: Error { error: Invalid, reason: "Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>" }
FAILED on shard 10: Error { error: Invalid, reason: "Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>" }
FAILED on shard 11: Error { error: Invalid, reason: "Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>" }
FAILED on shard 0: Error { error: Invalid, reason: "Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>" }
```

**Comment by kbr-scylla** [comment]
I modified the reproducer: after it gets this error from Scylla, I would sleep 100ms and retry the insert. If it's a race, the second insert should succeed.

However, the second insert also fails. So it looks like schema enters a broken state that is not transient?

**Comment by kbr-scylla** [comment]
Also, running 1000 with this diff:
https://github.com/scylladb/scylladb/issues/15530#issuecomment-1753275810

Passed.

So the most likely culprit is https://github.com/scylladb/scylladb/commit/3d4398d1b2b9bc2139718e49a806214171ee081f :(
I'm going to send a patch to disable this. (By disabling the cluster feature.)

**Comment by kbr-scylla** [comment]
I described the root cause here: https://github.com/scylladb/scylladb/pull/15687#issuecomment-1755861099
